File: chaeyun/3_palindrome/min-del-to-make-str-unique/sol2-sort.cpp
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <algorithm>
#include <string>

class Solution {
public:
    int minDeletions(std::string s) {
        std::unordered_map<char, int> cnt;
        for (char c : s) {
            cnt[c]++;
        }
        
        std::vector<int> sorted_freqs;
        for (const auto& pair : cnt) {
            sorted_freqs.push_back(pair.second);
        }
        std::sort(sorted_freqs.begin(), sorted_freqs.end(), std::greater<int>());
        
        int deletions = 0;
        std::unordered_set<int> used_frequencies;
        
        for (int freq : sorted_freqs) {
            if (used_frequencies.find(freq) == used_frequencies.end()) {
                used_frequencies.insert(freq);
                continue;
            }
            
            while (freq > 0 && used_frequencies.find(freq) != used_frequencies.end()) {
                freq--;
                deletions++;
            }
            
            if (freq > 0) {
                used_frequencies.insert(freq);
            }
        }
        
        return deletions;
    }
};

File: chaeyun/3_palindrome/min-del-to-make-str-unique/sol2-sort.py
# Sorting Approach
class Solution:
    def minDeletions(self, s: str) -> int:
        cnt = Counter(s)
        deletions = 0
        used_frequencies = set()
        
        sorted_freqs = sorted(cnt.values(), reverse=True)
        
        for freq in sorted_freqs:
            if freq not in used_frequencies:  # Early exit condition
                used_frequencies.add(freq)
                continue  

            while freq > 0 and freq in used_frequencies:
                freq -= 1
                deletions += 1

            used_frequencies.add(freq)
            
        return deletions